Add support for Rest-connectivity

ManuelMuehlberger opened the pull request:
Enabling gerrit-events to pull info from gerrit's Rest api. This is mainly intended for connectivity with our [ jira-gerrit](https://github.com/faktorzehn/jira-gerrit-plugin) fork.

ManuelMuehlberger committed:
add rest support

f10mlupfer commented on src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritQueryHandlerHttp.java:
```diff
@@ -0,0 +1,497 @@
+            logger.error("Failed to submit result to Gerrit", e);
+        }
+  */
+  }
```
das kann raus

f10mlupfer commented on src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritQueryHandlerHttp.java:
```diff
@@ -0,0 +1,497 @@
+  public interface Credential {
+    /**
+     * The user principal.
+     * @return the password
```
nicht richtig

f10mlupfer commented on src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritQueryHandlerHttp.java:
```diff
@@ -0,0 +1,497 @@
+  private final int statusOk = 200;
+  private final int statusBadRequest = 400;
+  private final int statusBadCredentials = 401;
+  private final int statusNotFound = 404;
```
könnte man als static final int STATUS_OK = 200; bspw. definieren, dann kann man sie in der Testklasse auch verwenden

f10mlupfer commented on src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritQueryHandlerHttp.java:
```diff
@@ -0,0 +1,497 @@
+ *
+ * @author Robert Sandell &lt;robert.sandell@sonyericsson.com&gt;
+ */
+public class GerritQueryHandlerHttp {
```
können wir den GerritQueryHandlerHttp von GerritQueryHandler erben lassen? Sparen wir uns da Implementierungen?

ManuelMuehlberger replied:
Ich weiß nicht, wieviel wir uns da sparen würden: GerritQueryHandler hat ja einige SSH-spezifische Attribute die in den Konstruktoren gesetzt werden. Wenn wir davon dann subclassen, müssen wir ja viele unnötige Attribute setzen? Entweder müssten wir einen neuen leeren Konstruktor hinzufügen, oder wir erstellen eine abstrakte Superklasse.

f10mlupfer replied:
würde ich gut finden (auch generell vielleicht auch nochmal aufzuräumen). Aber im Moment wäre es auch so ok für mich. Ggf. wäre das auch etwas für die TODO Liste im Wiki. Und evtl. fallen dir ja noch mehr Punkte ein, die man dort mit aufnehmen kann

ManuelMuehlberger replied:
Ich habe jetzt schon mal ein bisschen aufgeräumt, werde im Wiki aber auch eine TODO-Sektion aufmachen

f10mlupfer commented on src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritQueryHandlerHttp.java:
```diff
@@ -0,0 +1,497 @@
+    List<JSONObject> list = new ArrayList<>();
+
+    runQuery(queryString, getPatchSets, getCurrentPatchSet, getFiles, getCommitMessage, getComments,
+        new GerritQueryHandlerHttp.LineVisitor() {
+          @Override
+          public void visit(String line) {
+            JSONObject json = (JSONObject)JSONSerializer.toJSON(line.trim());
+            list.add(json);
+          }
+        });
```
die Methode runQuery ist ja private, da könnte man sich das LineVisitor Interface sparen und anstatt dessen einfach einen Consumer<String> übergeben und in der runQuery dann einfach in der Schleife visitor.accept(String) machen

f10mlupfer commented on src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritQueryHandlerHttp.java:
```diff
@@ -0,0 +1,497 @@
+    List<String> list = new ArrayList<>();
+
+    runQuery(queryString, getPatchSets, getCurrentPatchSet, getFiles, getCommitMessage, false,
+        new GerritQueryHandlerHttp.LineVisitor() {
+          @Override
+          public void visit(String line) {
+            list.add(line);
+          }
+        });
```
und hier dann einfach list::add

f10mlupfer commented on src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritQueryHandlerHttp.java:
```diff
@@ -0,0 +1,497 @@
+    JSONArray jsonArray = JSONArray.fromObject(body);
+
+    for (int i = 0; i < jsonArray.size(); i++) {
+      visitor.visit(jsonArray.getString(i));
```
visitor.accept(jsonArray.getString(i));
die for Loop könnte man auch noch per Stream machen (kein muss).

f10mlupfer commented on src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritQueryHandlerHttp.java:
```diff
@@ -0,0 +1,497 @@
+   * Internal visitor for handling a line of text.
+   * Used by .
+   */
+  interface LineVisitor {
+    /**
+     * Visits a line of query result.
+     *
+     * @param line the line.
+     */
+    void visit(String line);
+  }
```
wäre mit Consumer<String> dann überflüssig

f10mlupfer commented on src/test/java/com/sonymobile/tools/gerrit/gerritevents/GerritQueryHandlerHttpTest.java:
```diff
@@ -0,0 +1,423 @@
+    public void debug() {
+        //debugStub();
+        setupQueryHandlerHttp();
+        /*
+        given().auth().preemptive().basic(credentials.getUserPrincipal().getName(), credentials.getPassword())
+            .when().get("https://gerrit.faktorzehn.de/a/changes/?q=limit%3A1&o=ALL_REVISIONS&o=CURRENT_REVI
+            SION&o=CURRENT_FILES&o=CURRENT_COMMITS&o=MESSAGES").then()
+            .assertThat().statusCode(200);
+         */
+
+/*
+        try {
+            Response response = gerritQueryHandlerHTTP.runQuery("limit:2", true, true, true, true, true);
+            System.out.println();
+            String s = response.body().asPrettyString();
+            System.out.println(s);
+        } catch (IOException e) {
+            e.printStackTrace();
+        } catch (GerritQueryException e) {
+            e.printStackTrace();
+        }
+ */
```
das braucht es dann auch nicht

f10mlupfer commented on src/test/java/com/sonymobile/tools/gerrit/gerritevents/GerritQueryHandlerHttpTest.java:
```diff
@@ -0,0 +1,423 @@
+     * see {@link GerritQueryHandlerHttpTest#urlConversionTest()}.
+     * @throws Exception when something wrong
+     */
+    private void urlConversionTest3True() throws Exception {
```
True bedeutet hier, dass keine Exception geworfen wird? Würde dann eher Success nehmen. Gibt es auch Tests für Failure/Exception? Braucht es das?

ManuelMuehlberger replied:
Name wurde umgeändert zu urlConversionTestXSuccess(). Es gibt einen invalidRequestsThrowsTest() um sicher zu stellen, dass wir richtig mit (für den Server) ungültigen Requests umgehen.

f10mlupfer replied:
ok

f10mlupfer commented on src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritQueryHandlerHttp.java:
```diff
@@ -0,0 +1,497 @@
+
+    if (getPatchSets) {
+      //ALL_REVISIONS: describe all revisions, not just current.
+      str.append("&o=ALL_REVISIONS");
```
macht es ggf. Sinn, diese URL-Parameter ("&o=ALL_REVISIONS") in Konstanten auszulagern? Dann könnte man diese auch u.A. in den Tests verwenden

f10mlupfer reviewed (changes requested):
noch ein paar Anmerkungen und Gedankenanstösse, was man vielleicht noch anpassen könnte

ManuelMuehlberger committed:
resolve most code review comments

-removed unnecessary comments
-global variables instead of local literals
-rename test functions
-replace LineVisitor interface with Consumer<>

f10mlupfer commented on src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritQueryHandlerHttp.java:
```diff
@@ -5,59 +5,45 @@
-  private final int statusBadRequest = 400;
-  private final int statusBadCredentials = 401;
-  private final int statusNotFound = 404;
+  static final int statusOk = 200;
```
Konstanten würde ich komplett groß schreiben. Bspw. STATUS_OK

f10mlupfer commented on src/main/java/com/sonymobile/tools/gerrit/gerritevents/GerritQueryHandlerHttp.java:
```diff
@@ -5,59 +5,45 @@
+  private final String getCurrentRevision = "&o=CURRENT_REVISION";
+  private final String getCurrentFiles = "&o=CURRENT_FILES";
+  private final String getCurrentCommit = "&o=CURRENT_COMMIT";
+  private final String getMessages = "&o=MESSAGES";
```
auch die hier
ALL_REVISIONS
...

f10mlupfer reviewed (changes requested):
die Konsanten könnte man noch renamen, ansonsten ok

ManuelMuehlberger committed:
small cleanup

-all constants are uppercase
-parsing for loop is now forEach()
-remove unused constructor

ManuelMuehlberger committed:
update readme

ManuelMuehlberger committed:
small refactors to appease the checkstyle plugin

ManuelMuehlberger committed:
remove whitespace
